src/Npgsql/Util/TimeoutCancellationTokenSourceWrapper.cs

```diff
@@ -0,0 +1,72 @@
+﻿using System;
+using System.Threading;
+
+namespace Npgsql.Util
+{
+    /// <summary>
+    /// Internal struct wrapping a <see cref="CancellationTokenSource"/> for timeouts.
+    /// </summary>
+    /// <remarks>
+    /// Since there's no way to reset a <see cref="CancellationTokenSource"/> once it was cancelled,
+    /// we need to make sure that an existing cancellation token source hasn't been cancelled,
+    /// every time we start it (see https://github.com/dotnet/runtime/issues/4694).
+    /// </remarks>
+    class TimeoutCancellationTokenSourceWrapper : IDisposable
+    {
+        public TimeSpan Timeout { get; set; }
+        CancellationTokenSource _timeoutCts = new CancellationTokenSource();
+
+        public TimeoutCancellationTokenSourceWrapper() => Timeout = System.Threading.Timeout.InfiniteTimeSpan;
+
+        public TimeoutCancellationTokenSourceWrapper(TimeSpan timeout) => Timeout = timeout;
+
+        /// <summary>
+        /// Set the timeout on the wrapped <see cref="CancellationTokenSource"/>
+        /// and make sure that it hasn't been cancelled yet
+        /// </summary>
+        public CancellationToken Start()
+        {
+            _timeoutCts.CancelAfter(Timeout);
```

Comment:
I'd again add a DebugAssert that we're not already running.

Reply:
Ok. I'll do that.